s2spy/calendar_shifter.py

```diff
@@ -0,0 +1,188 @@
+        )
+    # combine gap and shift time dict
+    gap_shifted = {
+        k: gap_time_dict.get(k, 0) + shift_time_dict.get(k, 0)
+        for k in set(gap_time_dict) | set(shift_time_dict)
+    }
+
+    return gap_shifted
+
+
+def cal_shifter(
+    calendar: s2spy.time.Calendar, shift: Union[str, dict]
+) -> s2spy.time.Calendar:
+    """
+    Shift a Calendar instance by a given time offset. Instead of shifting the
+    anchor date, this function shifts two things in reference to the anchor date:
+    - the target period(s): as a gap between the anchor date and the start of the first
+        target period
+    - the precursor period(s): as a gap between the anchor date and the start of the
+        first precursor period
+    This way, the anchor year from the input calendar is maintained on the returned
+    calendar. This is important for train-test splitting at later stages.
+    - Args:
+        - calendar : a s2spy.time.Calendar instance
+        - shift: a pandas-like
+        frequency string (e.g. "10d", "2W", or "3M"), or a pandas.DateOffset
+        compatible dictionary such as {days=10}, {weeks=2}, or {months=1, weeks=2}
+    - Example:
+        Shift a calendar by a given dateoffset.
+        >>> import s2spy.time
+        >>> gap_shifted = gap_shift()
+        >>> cal = nwp_calendar_builder(start_date, '7d', '7d', precur_leadtime='14d', n_precurperiods=4)
+        >>> cal_shifted = shifter(cal, '7d')
+        Calendar(
+            anchor='07-01',
+            allow_overlap=False,
+            mapping=None,
+            intervals=[
+                Interval(role='target', length='7d', gap={'days': 7}),
+                Interval(role='precursor', length='7d', gap={'days': 7}),
+                Interval(role='precursor', length='7d', gap='0d'),
+                Interval(role='precursor', length='7d', gap='0d'),
+                Interval(role='precursor', length='7d', gap='0d'),
+            ]
+        )
+    """
+    # get 1st target interval and 1st precursor interval from calendar
+    target_interval = calendar.targets[0]
+    precursor_interval = calendar.precursors[0]
+
+    # shift gaps
+    target_gap_shifted = _gap_shift(target_interval, shift)
+    precursor_gap_shifted = _gap_shift(precursor_interval, shift)
+
+    # set the new gaps on the calendar
+    calendar_shifted = copy.deepcopy(calendar)
+    calendar_shifted.targets[0].gap = target_gap_shifted
+    calendar_shifted.precursors[0].gap = precursor_gap_shifted
+
+    return calendar_shifted
+
+
+def cal_list_resampler(cal_list: list, ds: xr.Dataset) -> xr.DataArray:
+    """
+    Resample a dataset to every calendar in a list of calendars and concatenate them
+    to a new dataset with dimension 'run'.
+    - Args:
+        - cal_list : list of shifted custom calendars
+        - ds: dataset to resample
+    """
+    # loop over calendar in the nwp_calendar_list
+    for x, cal_x in enumerate(cal_list):
+        cal_x_mapped = cal_x.map_to_data(ds)
```

Comment:
Hmm, yes I agree that it would be better for the user to see what is happening. Should we add a step in the notebook where we show how to map all the calendars in the list (either to some years or to data)?

Reply:
You can also map the calendar before you do the shifting, right? That would avoid having to do this when they are in a list. You can even check if the calendar is mapped before you do the shifting loop.

It could still be good to show how to do this afterwards though.